quint/src/effects/builtinSignatures.ts

```diff
@@ -12,175 +12,241 @@
  * @module
  */
 
-import { EffectScheme, Signature, effectNames } from './base'
+import { ComponentKind, Effect, EffectComponent, EffectScheme, Signature, Variables, effectNames, toScheme } from './base'
 import { parseEffectOrThrow } from './parser'
-import { range, times, zip } from 'lodash'
+import { range, times } from 'lodash'
 
 export function getSignatures(): Map<string, Signature> {
   return new Map<string, Signature>(fixedAritySignatures.concat(multipleAritySignatures))
 }
 
-const literals = ['Nat', 'Int', 'Bool'].map(name => ({ name, effect: 'Pure' }))
-
 function parseAndQuantify(effectString: string): EffectScheme {
   const e = parseEffectOrThrow(effectString)
   return { effect: e, ...effectNames(e) }
 }
 
+const prefixByComponentKind = new Map<ComponentKind, string>([
+  ['read', 'r'],
+  ['temporal', 't'],
+  ['update', 'u'],
+])
+
+/**
+ * Builds a signature that propagates components of the given kinds
+ *
+ * Example: propagateComponents(['read', 'temporal'])(2) results in
+ * (Read[r1] & Temporal[t1], Read[r2] & Temporal[t2]) => Read[r1, r2] & Temporal[t1, t2]
+ *
+ * @param kinds the kinds of components to propagate
+ * @returns an arrow function taking arity and returning the effect for that arity
+ */
+function propagateComponents(kinds: ComponentKind[]): ((arity: number) => EffectScheme) {
+  return (arity: number) => {
+    const params: Effect[] = times(arity, i => {
+      const components: EffectComponent[] = kinds.map(kind => {
+        return { kind: kind, variables: { kind: 'quantified', name: `${prefixByComponentKind.get(kind)!}${i + 1}` } }
+      })
+
+      return { kind: 'concrete', components: components }
+    })
+
+    const resultComponents: EffectComponent[] = kinds.map(kind => {
+      const names = times(arity, i => `${prefixByComponentKind.get(kind)!}${i + 1}`)
+      const variables: Variables[] = names.map(name => ({ kind: 'quantified', name }))
+      return { kind: kind, variables: { kind: 'union', variables } }
+    })
+
+    const result: Effect = { kind: 'concrete', components: resultComponents }
+    const effect: Effect = { kind: 'arrow', params: params, result: result }
+
+    return { effect, ...effectNames(effect) }
+  }
+}
+
+const p = parseAndQuantify
+const standardPropagation = propagateComponents(['read', 'temporal'])
+
+/**
+ * Builds a signature that propagates components of the given kinds, including a lambda parameter
+ * for the last argument
+ *
+ * Used in operators like `map`, `filter` and `fold`
+ *
+ * Example: propagationWithLambda(['read', 'temporal'])(2) results in
+ * (Read[r1] & Temporal[t1], (Read[r1] & Temporal[t1]) => Read[r2] & Temporal[t2]) => Read[r1, r2] & Temporal[t1, t2]
+ *
+ * @param kinds the kinds of components to propagate
+ * @returns an arrow function taking arity and returning the effect for that arity
+ */
+function propagationWithLambda(kinds: ComponentKind[]): ((arity: number) => EffectScheme) {
```

Comment:
It looks to me like these functions are helpers for generating specific effect signatures, rather than any sort of effect propagation inference of generator etc. With that understanding, a name like

```suggestion
function effectSignatureWithLambda(kinds: ComponentKind[]): ((arity: number) => EffectScheme) {
```

would be a bit clearer to me. That said, I may be misunderstanding, and this isn't a substantive concern for me.

Reply:
Hmm, I'm not sure. My intention with "propagation" here is that it defines an effect that propagates the given components. `effectSignatureWithLambda` is more generic, as a lot of different signatures like that can be defined. A signature that does propagation (that is, combines the effect of the parameters in the result) is much more specific. I'm also only using these helpers for operators that do full propagation, and there are some effects that are still defined manually.